src/queue.cpp

```diff
@@ -258,15 +268,102 @@ cl_int cvk_command_queue::wait_for_events(cl_uint num_events,
     return ret;
 }
 
+static cl_int execute_cmds(std::deque<cvk_command*>& cmds) {
+    TRACE_FUNCTION();
+    cl_int global_status = CL_SUCCESS;
+    while (!cmds.empty()) {
+        cvk_command* cmd = cmds.front();
+        cvk_debug_fn("executing command %p, event %p", cmd, cmd->event());
+
+        cl_int status = cmd->execute();
+        if (status != CL_COMPLETE && global_status == CL_SUCCESS)
+            global_status = status;
+        cvk_debug_fn("command returned %d", status);
+
+        cmds.pop_front();
+
+        delete cmd;
+    }
+    return global_status;
+}
+
+cl_int cvk_command_queue::execute_cmds_dominated_by_no_lock(
+    cl_uint num_events, _cl_event* const* event_list) {
+    auto* exec = m_executor;
+    if (exec == nullptr) {
+        return CL_SUCCESS;
+    }
+
+    m_lock.unlock();
+    auto cmds = exec->extract_cmds_dominated_by(false, num_events, event_list);
+    auto ret = execute_cmds(cmds);
```

Comment:
I think we are losing the dependency of the commands that are left to the executor on the commands that are executed synchronously. We will need these to handle error propagation. I suggest we add error propagation tests checking that commands run by the executor get their error status from previously enqueued commands that were stolen for synchronous execution (exercising both the clWaitForEvents and clFinish cases).